File: src/arch/i386/kernel/gdt.c
#include "include/gdt.h"
#include "include/console.h"
#include "include/libc/include/defines.h"
#include "include/libc/include/string.h"
#include "include/libc/include/types.h"

GDT gdt[GDT_ENTRIES];
GDT_PTR gp;
tss_entry tss;
u32 page_directory[1024] __attribute__((aligned(4096)));
u32 first_page_table[1024] __attribute__((aligned(4096)));

void set_gdt_entry(int num, u32 base, u32 limit, u8 access, u8 gran) {
  gdt[num].base_low = (base & 0xFFFF);
  gdt[num].base_middle = (base >> 16) & 0xFF;
  gdt[num].base_high = (base >> 24) & 0xFF;

  gdt[num].limit_low = (limit & 0xFFFF);
  gdt[num].granularity = ((limit >> 16) & 0x0F);
  gdt[num].granularity |= (gran & 0xF0);
  gdt[num].access = access;
}

void setup_paging() {
  int i;
  // Identity-map the first 4MB of memory.
  for (i = 0; i < 1024; i++) {
    first_page_table[i] = (i * 0x1000) | PAGE_PRESENT | PAGE_RW | PAGE_USER;
  }

  page_directory[0] =
      ((u32)first_page_table) | PAGE_PRESENT | PAGE_RW | PAGE_USER;

  for (i = 1; i < 1024; i++) {
    page_directory[i] = 0;
  }

  asm volatile("mov %0, %%cr3" : : "r"(page_directory));

  u32 cr0;
  asm volatile("mov %%cr0, %0" : "=r"(cr0));
  cr0 |= 0x80000000;
  asm volatile("mov %0, %%cr0" : : "r"(cr0));
}

void usermode() {
  asm volatile(
      "cli\n" // Disable interrupts.
      // Set up segment registers for user mode.
      "mov $0x23, %%ax\n" // User data segment (selector index 4, RPL=3: 0x23).
      "mov %%ax, %%ds\n"
      "mov %%ax, %%es\n"
      "mov %%ax, %%fs\n"
      "mov %%ax, %%gs\n"
      "mov %%esp, %%eax\n" // Save current ESP.
      "pushl $0x23\n"      // Push user SS.
      "pushl %%eax\n"      // Push user ESP.
      "pushfl\n"           // Push EFLAGS.
      "pushl $0x1B\n"      // Push user CS (selector index 3, RPL=3: 0x1B).
      "pushl $1f\n"        // Push entry point.
      "iret\n"             // Return to user mode.
      "1:\n"
      "nop\n"
      :
      :
      : "eax");
}

void init_gdt() {
  printf("initiating GDT...\n");
  gp.limit = (sizeof(gdt) * GDT_ENTRIES) - 1;
  gp.base = (u32)&gdt;

  set_gdt_entry(0, 0, 0, 0, 0);
  set_gdt_entry(1, 0, 0xFFFFFFFF, 0x9A, 0xCF);
  set_gdt_entry(2, 0, 0xFFFFFFFF, 0x92, 0xCF);
  set_gdt_entry(3, 0, 0xFFFFFFFF, 0xFA, 0xCF);
  set_gdt_entry(4, 0, 0xFFFFFFFF, 0xF2, 0xCF);

  memset(&tss, 0, sizeof(tss));
  tss.ss0 = 0x10;
  tss.esp0 = (u32)stack_top;
  tss.iomap_base = sizeof(tss);
  set_gdt_entry(5, (u32)&tss, sizeof(tss), 0x89, 0x00);

  load_gdt((u32)&gp);
  asm volatile("ltr %0" : : "r"((u16)0x28));
  setup_paging();
}

File: src/arch/i386/gdt.zig
const io = @import("../../utils/io.zig");
const boot = @import("../../boot.zig");

pub const GDT_SELECTOR = struct {
    pub const NULL: u16        = 0x00;
    pub const KERNEL_CODE: u16 = 0x08;
    pub const KERNEL_DATA: u16 = 0x10;
    pub const USER_CODE: u16   = 0x18;
    pub const USER_DATA: u16   = 0x20;
    pub const TSS: u16         = 0x28;
};

pub const GDTPtr = packed struct {
    size: u16, // Size of GDT - 1
    addr: u32,  // Physical address of GDT
};

const GDTEntry = packed struct {
    limit_low: u16,
    base_low: u16,
    base_middle: u8,
    access: u8,
    granularity: u8,
    base_high: u8,
};

const TSS = packed struct {
    prev_tss: u16,
    reserved1: u16,
    esp0: u32,
    ss0: u16,
    reserved2: u16,
    esp1: u32,
    ss1: u16,
    reserved3: u16,
    esp2: u32,
    ss2: u16,
    reserved4: u16,
    cr3: u32,
    eip: u32,
    eflags: u32,
    eax: u32,
    ecx: u32,
    edx: u32,
    ebx: u32,
    esp: u32,
    ebp: u32,
    esi: u32,
    edi: u32,
    es: u16,
    reserved5: u16,
    cs: u16,
    reserved6: u16,
    ss: u16,
    reserved7: u16,
    ds: u16,
    reserved8: u16,
    fs: u16,
    reserved9: u16,
    gs: u16,
    reserved10: u16,
    ldtr: u16,
    reserved11: u16,
    trap: u16,
    io_permissions_base_offset: u16,
};

var gdt = [_]u64{
    // Null descriptor
    0x0000000000000000,
    // Kernel code (0x08)
    createDescriptor(0, 0xFFFFF, 0x9A, 0xCF),
    // Kernel data (0x10)
    createDescriptor(0, 0xFFFFF, 0x92, 0xCF),
    // User code (0x18)
    createDescriptor(0, 0xFFFFF, 0xFA, 0xCF),
    // User data (0x20)
    createDescriptor(0, 0xFFFFF, 0xF2, 0xCF),
    // TSS descriptor (0x28)
    0,
};

var gdtr: GDTPtr = undefined;
var tss: TSS    = undefined;

// zig fmt: off
pub fn load() void {
    gdtr = .{
        .limit = @sizeOf(@TypeOf(gdt)) - 1,
        .base  = @intFromPtr(&gdt),
    };

    io.lgdt(&gdtr);

    // Reload CS with far jump
    asm volatile (
        \\ ljmp $0x08, $flush
        \\ flush:
        :
        :
        : "memory"
    );

    // Reload data segments
    asm volatile (
        \\ mov $0x10, %%ax
        \\ mov %%ax, %%ds
        \\ mov %%ax, %%es
        \\ mov %%ax, %%fs
        \\ mov %%ax, %%gs
        \\ mov %%ax, %%ss
        :
        :
        : "ax", "memory"
    );

    init_tss();
}

fn createDescriptor(base: u32, limit: u32, access: u8, gran: u8) u64 {
    const limit_low   = limit & 0xFFFF;        // bits 0..15
    const base_low    = base  & 0xFFFF;        // bits 16..31
    const base_mid    = (base >> 16) & 0xFF;   // bits 32..39
    const access_byte = access;                // bits 40..47
    const limit_high  = (limit >> 16) & 0x0F;  // bits 48..51
    const flags       = gran & 0x0F;           // bits 52..55
    const base_high   = (base >> 24) & 0xFF;   // bits 56..63

    return  (@as(u64, limit_low)        << 0)
         | (@as(u64, base_low)         << 16)
         | (@as(u64, base_mid)         << 32)
         | (@as(u64, access_byte)      << 40)
         | (@as(u64, limit_high)       << 48)
         | (@as(u64, flags)            << 52)
         | (@as(u64, base_high)        << 56);
}
// zig fmt: on

fn init_tss() void {
    const tss_selector: u16 = GDT_SELECTOR.TSS;
    const tss_base          = @intFromPtr(&tss);
    const tss_limit         = @sizeOf(TSS) - 1;
    const stack_top         = @intFromPtr(&boot.stack_bytes) + @sizeOf(@TypeOf(boot.stack_bytes));

    tss.esp0 = @as(u32, stack_top);
    tss.ss0  = GDT_SELECTOR.KERNEL_DATA;

    // Update TSS descriptor
    gdt[5] = createDescriptor(tss_base, tss_limit, 0x89, 0x00);

    tss.io_permissions_base_offset = @sizeOf(TSS); // Disable I/O permission bitmap

    io.ltr(tss_selector);
}
